File: app/src/main/java/com/example/administrator/itemdecoration/itemdecoration/TimeLineDecoration.java
package com.example.administrator.itemdecoration.itemdecoration;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.RecyclerView.State;
import android.view.View;

/**
 * Created by lengwei on 2017/11/6.
 *
 * @Description 时间轴
 */
public class TimeLineDecoration extends RecyclerView.ItemDecoration {

    private Context mContext;
    private Paint mPaint;
    private Paint mTextPaint;
    private int moveX = 250; //Rect left move
    private int moveY = 50;  //Rect top move
    private static final int RADIUS = 12; //半径

    public TimeLineDecoration(Context context) {
        this.mContext = context;
        this.mPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        this.mPaint.setColor(ContextCompat.getColor(context, android.R.color.holo_red_dark));
        this.mTextPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        this.mTextPaint.setColor(ContextCompat.getColor(context, android.R.color.holo_red_dark));
        this.mTextPaint.setTextSize(30);
    }

    @Override
    public void onDraw(Canvas c, RecyclerView parent, State state) {
        super.onDraw(c, parent, state);
        int childCount = parent.getChildCount();
        for (int i = 0; i < childCount; i++) {
            View child = parent.getChildAt(i);
            final int topX = child.getLeft() - moveX / 3;
            final int topY = child.getTop() - moveY;
            final int circleTopX = topX;
            final int circleTopY = child.getTop() + child.getHeight() / 2;
            final int circleX = topX;
            final int circleY = circleTopY + RADIUS;
            final int circleBottomX = topX;
            final int circleBottomY = circleY + RADIUS;
            final int bottomX = topX;
            final int bottomY = child.getBottom();
            c.drawText("时间轴", 50, circleY + 10, mTextPaint);
            c.drawLine(topX, topY, circleTopX, circleTopY, mPaint);
            c.drawCircle(circleX, circleY, RADIUS, mPaint);
            c.drawLine(circleBottomX, circleBottomY, bottomX, bottomY, mPaint);
        }
    }

    @Override
    public void getItemOffsets(Rect outRect, View view, RecyclerView parent, State state) {
        super.getItemOffsets(outRect, view, parent, state);
        outRect.left = moveX;
        outRect.top = moveY;
    }
}

File: app/src/main/java/com/example/administrator/itemdecoration/itemdecoration/TimeLineDecoration.kt
package com.example.administrator.itemdecoration.itemdecoration

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Rect
import android.support.v4.content.ContextCompat
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.RecyclerView.State
import android.view.View

/**
 * Created by lengwei on 2017/11/6.
 *
 * @Description 时间轴
 */
class TimeLineDecoration(mContext: Context) : RecyclerView.ItemDecoration() {
    private val mPaint: Paint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val mTextPaint: Paint
    /**
     * Rect left move
     */
    private val moveX = 250
    /**
     * Rect top move
     */
    private val moveY = 50

    init {
        this.mPaint.color = ContextCompat.getColor(mContext, android.R.color.holo_red_dark)
        this.mTextPaint = Paint(Paint.ANTI_ALIAS_FLAG)
        this.mTextPaint.color = ContextCompat.getColor(mContext, android.R.color.holo_red_dark)
        this.mTextPaint.textSize = 30f
    }

    override fun onDraw(c: Canvas, parent: RecyclerView, state: State) {
        super.onDraw(c, parent, state)
        val childCount = parent.childCount
        for (i in 0 until childCount) {
            val child = parent.getChildAt(i)
            val topX = child.left - moveX / 3
            val topY = child.top - moveY
            val circleTopY = child.top + child.height / 2
            val circleY = circleTopY + DOT_RADIUS
            val circleBottomY = circleY + DOT_RADIUS
            val bottomY = child.bottom
            c.drawText("时间轴", 50f, (circleY + 10).toFloat(), mTextPaint)
            c.drawLine(topX.toFloat(), topY.toFloat(), topX.toFloat(), circleTopY.toFloat(), mPaint)
            c.drawCircle(topX.toFloat(), circleY.toFloat(), DOT_RADIUS.toFloat(), mPaint)
            c.drawLine(topX.toFloat(), circleBottomY.toFloat(), topX.toFloat(), bottomY.toFloat(), mPaint)
        }
    }

    override fun getItemOffsets(outRect: Rect, view: View, parent: RecyclerView, state: State) {
        super.getItemOffsets(outRect, view, parent, state)
        outRect.left = moveX
        outRect.top = moveY
    }

    companion object {
        /**
         * 半径
         */
        private val DOT_RADIUS = 12
    }
}
